### DISCUSSION TITLE
Access request headers in hook read filter

### DISCUSSION THREAD

**Discussion by mxgnr** [directus/directus#12484] [Archive]
Hi there,

I would like to access the request object in my `(<collection>.)items.read` hook to get the cookies data sent with the request, but i can't find any info in the docs or in the threads to do it. I've found the new authenticate hook (#11942) but it's not suitable for me.

Is there any way to do that ?

Thanks for the help and for all the hard work on Directus :)

**Comment by rijkvanzanten** [comment]
The request is currently not passed to the hook context, as there's no guarantee that those calls are coming in through a http request 🤔 We'll have to think through making sure that we can figure this out in a way where the signature of the context object is consistent, regardless of where the method is called from

**Reply by mxgnr** [reply]
Got it. It's the same issue for the action hook i imagine ? Is there any way i can get the cookies data in the hook `action` instead of `filter`, or with node/express middleware for example ?

But anyway, this is what i'm trying to implement, maybe it will help you to better understand what i'm trying to do :

By default, my app does not require an account creation, this is an intended behavior. 
But there is two roles : guest (public) and author (public/user). Guest will always be a public role, and a author begin as public and can become a authenticated user afterwards (by user request, if he wants to keep the data linked with his account). 

Guest can interact with the collection "foo" through a private link which contains a token (UUID), saved as cookie when the page is loaded. Therefore, the token needs to be checked if it matches with the UUID saved in the item collection (`guest_key`) for each request (CRUD). The same behavior will be done for the author token (`author_key`), with some extra accountability.

That's why i was thinking about using the hooks for checking the token between client & item collection when the request is made, but maybe there is a better way to do that ? Since i do not want to force users to create an account, this is the main issue.

Thanks

**Reply by rijkvanzanten** [reply]
Ahh gotcha. Maybe the `authenticate` filter hook is better? That allows for a custom auth check against the request object which allows you to define the accountability object the API uses for permission checks downstream. That way, you could effectively "login" the user for the duration of the request based on the incoming headers (cookie f.e.) or other request properties 🤔

**Reply by mxgnr** [reply]
Seems to be good to me, however how to filter this behavior to a specific route/item ?

**Reply by mxgnr** [reply]
I've found `req._parsedUrl.pathname` property that i can compare, it seems to be right ?

**Reply by rijkvanzanten** [reply]
The `req` object is a standard Express Request object, so all of the properties found here are available: https://expressjs.com/en/4x/api.html#req so any one of these can/should be used:

```js
  console.dir(req.originalUrl) // '/admin/new?sort=desc'
  console.dir(req.baseUrl) // '/admin'
  console.dir(req.path) // '/new'
```
